### PR TITLE
GH-36379: [C++] Bundled dependency include paths should override system include dirs

### LINKED ISSUE #36379: [C++] Bundled GTest CI jobs seem to `#include`ing GTest headers from somewhere else
### Describe the bug, including details regarding any error messages, version, and platform.

In https://github.com/apache/arrow/pull/36334 I saw link errors in some CI jobs which build bundled GTest:

- [AMD64 manylinux2014 Java JNI](https://github.com/apache/arrow/actions/runs/5404201925/jobs/9818060319?pr=36334#step:6:2663)
- [AMD64 macOS 12 C++](https://github.com/apache/arrow/actions/runs/5404201927/jobs/9818061642?pr=36334#step:9:1282)

The symbol which is undefined is the internal [FormatMatcherDescription](https://github.com/google/googletest/blame/687c589949eaf8734484491a6882a7fc56aebc12/googlemock/include/gmock/gmock-matchers.h#L3875) function. That function is present in the bundled version of GTest (v1.11), but its signature [is different in that version](https://github.com/google/googletest/blob/release-1.11.0/googlemock/include/gmock/gmock-matchers.h#L3705).

This suggests that somehow although we're building GTest v1.11 the tests are `#include`ing headers from a more recent version of GTest and we just haven't noticed before now.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by bkietz**
### Rationale for this change

Bundled dependencies' include directories should override system include dirs. Otherwise an incompatible header in the system might be included when we wanted a header from the bundled dependency.

### What changes are included in this PR?

bundled dependencies explicitly insert their own include dirs ahead of others
* Closes: #36379

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -2600,6 +2602,7 @@ endmacro()
 if(ARROW_WITH_BZ2)
   resolve_dependency(BZip2 PC_PACKAGE_NAMES bzip2)
 
+  # XXX if resolve_dependency is called, won't we have either BZip2::BZip2 or an error?
   if(NOT TARGET BZip2::BZip2)
     add_library(BZip2::BZip2 UNKNOWN IMPORTED)
     set_target_properties(BZip2::BZip2
```
Comment (bkietz) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2608]: Is this dead code?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2608]: This block is for CMake < 3.12. So we can remove this.

**Review by kou — APPROVED** [review]
+1

We need to run `cmake-format` before we merge this.

**Review by assignUser — APPROVED** [review]
This change makes sense imo, a more involved (and brittle) alternative might be to move gtest to the front in `target_link_libraries` calls. I tried to reproduce the error but failed, where you able to test the fix with the error locally @bkietz ?

**Comment by pitrou** [comment]
Should we try to revert https://github.com/apache/arrow/pull/37483 in this PR?

**Comment by kou** [comment]
> Should we try to revert #37483 in this PR?

Is it just for confirming whether this fixes the #36379 problem or not? Do we re-revert #37483 after we confirm it?

#37483 is needed to fix #37067. So we should not revert #37483 from main.

**Comment by pitrou** [comment]
Ah, I thought the two approaches were alternatives. If not then nevermind.

**Comment by bkietz** [comment]
FetchContent_MakeAvailable automatically adds include directories and (at least while I was testing this) seems to *prepend* the `-I` flags by default. If I'm correct, then https://github.com/apache/arrow/pull/37483 actually resolved this issue for bundled GTest (but the other bundled dependencies are still vulnerable).

**Comment by kou** [comment]
> the other bundled dependencies are still vulnerable

I think so too.
So we need both of #37483 and this.

I merge this.

**Pull request merged**